### PR TITLE
Initial design

### PR DESCRIPTION
**Pull request by hannahhoward**
# Goals

Design basic methods for loading/storing nodes in the IPFS context (with a block store backing)

# Implementation

For now, this just offers a Store mechanism and a Delete mechanism. The Store is just a pass through to the LinkSystem store method, backed by the blockservice.

# For Discussion

- As I was writing this, it occurred to me that I think we ought to initialize the fetcher here. That's why I added the NewSession method
- Should this wrap a blockservice? or a blockstore? I think for now, it's ok to wrap a block service. When we start looking at doing fetches with mechanisms other than Bitswap, I think it will make sense to remove the blockservice entirely and spread the code out over this service and the fetcher.
- Should we have a mechanism for writing multiple blocks at once? BlockService has an AddBlocks method. I think it might be useful to do some kind of "new transaction" that would return a dag service you could store multiple changes to before actually writing.
- would it be helpful to have a Replace method with a signature like this: 
```
func Replace(linkCtx LinkContext, link ipld.Link, func(ipld.Node) ipld.Node)
```
?

### CODE REVIEW

**Comment by willscott** [comment]
* not too worried about the design for how the fetcher and blockstore composition looks, since we'll only do it one place and can iterate on it fairly easily if it ends up having rough edges.
* being able to batch writes (both adds and removes?) seems useful from a performance perspective and worth having
* i'd probably shy away from a replace at this level (weakly held opinion)

**Comment by hannahhoward** [comment]
awesome I think I'm going to remove the fetcher for now and add a multiwriter.

**Comment by hannahhoward** [comment]
I've added a BatchWriter + some tests, and am directly accessing the blockstore for now instead of blockservice... though for compatibility our best bet is blockservice

File: dagwriter.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+	"fmt"
+	"io"
+
+	blocks "github.com/ipfs/go-block-format"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	"github.com/ipld/go-ipld-prime"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+)
+
+type DagWriter interface {
```
Comment (willscott) [inline dagwriter.go:17]: comments on interface & interface methods here & below will be important

File: dagwriter.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+	Store(lnkCtx ipld.LinkContext, lp ipld.LinkPrototype, n ipld.Node) (ipld.Link, error)
```
Comment (willscott) [inline dagwriter.go:20]: will this fully recurse storing links in the node? what if there's a link that can't be resolved? is there a variant with a selector /walker?

Comment (hannahhoward) [reply dagwriter.go:20]: What does "what if there's a link that can't be resolved?" mean? This is purely storage -- we're not loading/resolving anything here.

Comment (willscott) [reply dagwriter.go:20]: Lets say the node I provide to store is a map containing a link. Will that link be followed?

File: dagwriter.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	Delete(ipld.Link) error
```
Comment (Stebalien) [inline dagwriter.go]: These should take contexts. Eventually, the blockstore should take contexts as well but we need to be able to cancel remote/slow operations.

Comment (Stebalien) [reply dagwriter.go]: Unless the dag writer is some "ephemeral" thing and we're supposed to to construct it with a context? But I'd prefer not to do that.

Comment (hannahhoward) [reply dagwriter.go]: LinkContext has a Context inside.

Delete doesn't have have a context in Blockstore / BlockService -- should I give it one?

Comment (Stebalien) [reply dagwriter.go]: > LinkContext has a Context inside.

https://github.com/ipld/go-ipld-prime/issues/154

> Delete doesn't have have a context in Blockstore / BlockService -- should I give it one?

Yes. The plan has been to slowly pipe the context all the way down into the datastore. If for nothing else, metrics.

Same with any flush/commit methods.

**Pull request merged**